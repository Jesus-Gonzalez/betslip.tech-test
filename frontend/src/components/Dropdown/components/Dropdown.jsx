import React from 'react'
import PropTypes from 'prop-types'

import {
  DropdownSelector,
  DropdownTrigger,
  DropdownWrapper
} from 'components/Dropdown/components'

export const Dropdown = React.forwardRef((props, ref) => {
  const { isVisible } = props

  return (
    <DropdownWrapper {...props}>
      <DropdownTrigger {...props} />
      {isVisible && <DropdownSelector {...props} />}
    </DropdownWrapper>
  )
})

Dropdown.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  setIsVisible: PropTypes.func.isRequired
}
