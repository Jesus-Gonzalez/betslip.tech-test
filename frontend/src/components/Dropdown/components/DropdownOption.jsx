import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import styles from './DropdownOption.module.scss'

const propTypes = {
  labelKey: PropTypes.string,
  option: PropTypes.object.isRequired
}

const defaultProps = {
  labelKey: 'label'
}

export const DropdownOption = ({
  className,
  labelKey,
  onChange,
  option,
  selected,
  ...props
}) => {
  const isSelected = React.useMemo(
    () => selected === option,
    [selected, option]
  )

  const optionClassname = React.useMemo(
    () => classnames(className, styles.option, isSelected && styles.selected),
    [className, isSelected]
  )

  return (
    <li
      className={optionClassname}
      data-dropdown-clickable
      onClick={() => onChange(option)}
    >
      {option[labelKey || 'label']}
    </li>
  )
}

DropdownOption.propTypes = propTypes

DropdownOption.defaultProps = defaultProps
