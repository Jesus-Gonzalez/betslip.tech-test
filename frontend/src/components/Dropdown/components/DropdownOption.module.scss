@import 'styles/variables';

.option {
  padding: 5px 10px;
  cursor: pointer;

  border-bottom: $borderGrey;

  &:hover {
    background-color: $colorLightGrey;
  }

  &:last-child {
    border-bottom: none;
  }
}

.selected {
  background-color: $colorGrey;
}
