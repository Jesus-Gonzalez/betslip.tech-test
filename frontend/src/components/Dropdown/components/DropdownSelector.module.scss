@import 'styles/variables';

.wrapper {
  position: relative;
  width: 100%;
}

.selector {
  position: absolute;
  display: flex;
  flex-direction: column;
  border: $borderGrey;
  border-top: none;
  list-style: none;
  padding: 0;
  margin: 0;
  width: 100%;
  background: #fff;
}
