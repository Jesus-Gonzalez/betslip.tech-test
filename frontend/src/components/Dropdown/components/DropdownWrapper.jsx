import React from 'react'
import classnames from 'classnames'

import styles from './DropdownWrapper.module.scss'

export const DropdownWrapper = ({
  children,
  className,
  ...props
}) => {
  const wrapperClass = React.useMemo(
    () => classnames(className, styles.dropdown),
    [className]
  )

  return (
    <div className={wrapperClass}>
      {children}
    </div>
  )
}
