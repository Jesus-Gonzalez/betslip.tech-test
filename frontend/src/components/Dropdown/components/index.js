export * from './Dropdown'
export * from './DropdownOption'
export * from './DropdownSelector'
export * from './DropdownTrigger'
export * from './DropdownWrapper'
