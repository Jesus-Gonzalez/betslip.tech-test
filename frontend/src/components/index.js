export * from './Button'
export * from './Card'
export * from './Dropdown'
export * from './Image'
export * from './Input'
export * from './PageWrapper'
