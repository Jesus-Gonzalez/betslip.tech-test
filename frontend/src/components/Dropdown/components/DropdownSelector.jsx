import React from 'react'
import PropTypes from 'prop-types'

import { DropdownOption } from 'components/Dropdown/components'

import styles from './DropdownSelector.module.scss'

const propTypes = {
  className: PropTypes.string,
  labelKey: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(PropTypes.object).isRequired,
  selected: PropTypes.object
}

const defaultProps = {
  className: '',
  labelKey: '',
  selected: null
}

export const DropdownSelector = ({
  className,
  labelKey,
  onChange,
  options,
  selected,
  ...props
}) => {
  const renderedOptions = React.useMemo(
    () => options.map(option => (
      <DropdownOption
        key={option.value}
        labelKey={labelKey}
        onChange={onChange}
        option={option}
        selected={selected}
      />
    )),
    [options, labelKey, onChange, selected]
  )

  return (
    <div className={styles.wrapper}>
      <ul className={styles.selector}>
        {renderedOptions}
      </ul>
    </div>
  )
}

DropdownSelector.propTypes = propTypes

DropdownSelector.defaultProps = defaultProps
