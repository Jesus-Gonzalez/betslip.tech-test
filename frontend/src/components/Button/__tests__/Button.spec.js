import React from 'react'
import { shallow } from 'enzyme'

import { Button } from 'components'

describe('components(Button)', () => {
  it('should render', () => {
    const button = shallow(<Button />).html()

    expect(button).toMatchSnapshot()
  })

  it('should render className prop', () => {
    const button = shallow(<Button className="test-class" />).html()

    expect(button).toMatchSnapshot()
  })

  it('should render type submit on submit prop', () => {
    const button = shallow(<Button submit />).html()

    expect(button).toMatchSnapshot()
  })

  it('should render innerText on button component', () => {
    const button = shallow(<Button submit>Submit form</Button>).html()

    expect(button).toMatchSnapshot()
  })

  it('should render rest props', () => {
    const fn = jest.fn(event => event)
    const button = shallow(<Button onClick={fn}>Click event simulation</Button>)
    button.find('button').simulate('click')

    expect(fn).toHaveBeenCalled()
  })
})
