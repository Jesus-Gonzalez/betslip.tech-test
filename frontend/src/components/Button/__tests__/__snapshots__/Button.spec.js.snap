// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components(Button) should render 1`] = `"<button type=\\"button\\" class=\\"button\\"></button>"`;

exports[`components(Button) should render className prop 1`] = `"<button type=\\"button\\" class=\\"test-class button\\"></button>"`;

exports[`components(Button) should render innerText on button component 1`] = `"<button type=\\"submit\\" class=\\"button\\">Submit form</button>"`;

exports[`components(Button) should render type submit on submit prop 1`] = `"<button type=\\"submit\\" class=\\"button\\"></button>"`;
